src/java.desktop/windows/native/libjsound/PLATFORM_API_WinOS_DirectSound.cpp

```diff
@@ -479,7 +479,7 @@ DS_StartBufferHelper::Data::~Data() {
 
 DWORD WINAPI __stdcall DS_StartBufferHelper::ThreadProc(void *param)
 {
-    ::CoInitialize(NULL);
+    ::CoInitializeEx(NULL, COINIT_MULTITHREADED | COINIT_DISABLE_OLE1DDE);
```

Comment:
You're changing the COM for this thread from `COINIT_APARTMENTTHREADED` to `COINIT_MULTITHREADED`.

Do DirectSound objects support multi-threading? I couldn't find anything quickly.

A Microsoft sample for [Creating the Device Object](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/ee416365(v=vs.85)) uses `CoInitializeEx(NULL, 0)`.

The documentation for [`CoInitializeEx`](https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-coinitializeex) says, “You need to initialize the COM library on a thread before you call any of the library functions…”

This implies that any thread which creates DirectSound objects must initialise COM first. I can't see it happening. Do I miss anything?

Reply:
@aivanov-jdk, Thank you for your time and reviews,
`CoInitializeEx(NULL, 0)` also resolving this problem because as per document  **The default is COINIT_MULTITHREADED**.

In my debug session noticed `DirectSoundCaptureEnumerate` getting called ahead of thread call or ::CoInitialize and that was the root cause this failure.

For resolving this found 3 ways and all works:
1 . Use COM multi-threading  :- Simplest approach with respect to other solutions 
2.  Hold the call till CoInitialize : - Can end up in deadlock situation so that need to be taken care.
3.  Do another CoInitialize before call  :- not a good approach

_This implies that any thread which creates DirectSound objects must initialise COM first. I can't see it happening. Do I miss anything?_
`DS_StartBufferHelper` class initializing COM in thread and all member functions checking initialization status before making any calls. So another solution may be restructure the code for accessing `isInitialized()` method from outer methods and proceed. (approach 2 in better way)

Reply:
> @aivanov-jdk, Thank you for your time and reviews, `CoInitializeEx(NULL, 0)` also resolving this problem because as per document **The default is COINIT_MULTITHREADED**.

This is what I expected, however, the documentation for [`COINIT`](https://learn.microsoft.com/en-us/windows/win32/api/objbase/ne-objbase-coinit) doesn't spell out the value for `COINIT_MULTITHREADED`.

> In my debug session noticed `DirectSoundCaptureEnumerate` getting called ahead of thread call or ::CoInitialize and that was the root cause this failure.

If it is the case, I can't see that you have ensured that `DirectSoundCaptureEnumerate` is called after `CoInitialize`. Do I miss anything?

> For resolving this found 3 ways and all works:
> 
> 1. Use COM multi-threading  :- Simplest approach with respect to other solutions

Probably… But it is still incorrect. Initialising COM on *a thread* doesn't mean you can call COM object methods from **any thread** in your application.

As the documentation says, “You need to initialize the COM library on a thread before you call any of the library functions…” (I have this very quote in [my comment above](https://github.com/openjdk/jdk/pull/14898#discussion_r1272027291).)

> 2. Hold the call till CoInitialize : - Can end up in deadlock situation so that need to be taken care.

It's a challenge but it is the only right way, as far as I can see… Perhaps not to wait till `::CoInitialize` is called on a thread (see above) but to transfer a call for handling to a thread where COM is initialised.

> 3. Do another CoInitialize before call  :- not a good approach

Yet it does not break the rules: you can initialise COM, call an API and then call [`CoUninitialize`](https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-couninitialize). If you don't keep any references to COM objects, it is valid.


> _This implies that any thread which creates DirectSound objects must initialise COM first. I can't see it happening. Do I miss anything?_ `DS_StartBufferHelper` class initializing COM in thread and all member functions checking initialization status before making any calls. So another solution may be restructure the code for accessing `isInitialized()` method from outer methods and proceed. (approach 2 in better way)
> 
> Tried this approach with `::CoInitialize` and observed _truncated names_, but its not failing for lock unlock workflow.

As I explained above, it would still be incorrect: if enumerating sound devices requires COM, it must be initialised on the thread where you enumerate the devices.

Reply:
> > @aivanov-jdk, Thank you for your time and reviews, `CoInitializeEx(NULL, 0)` also resolving this problem because as per document **The default is COINIT_MULTITHREADED**.
> 
> This is what I expected, however, the documentation for [`COINIT`](https://learn.microsoft.com/en-us/windows/win32/api/objbase/ne-objbase-coinit) doesn't spell out the value for `COINIT_MULTITHREADED`.

As per the declaration yes its 0 [tagCOINIT](https://github.com/wine-mirror/wine/blob/master/include/objbase.h) 
> 
> > In my debug session noticed `DirectSoundCaptureEnumerate` getting called ahead of thread call or ::CoInitialize and that was the root cause this failure.
> 
> If it is the case, I can't see that you have ensured that `DirectSoundCaptureEnumerate` is called after `CoInitialize`. Do I miss anything?
Yes, You are right in main thread no CoInitialize
> 
> > For resolving this found 3 ways and all works:
> > 
> > 1. Use COM multi-threading  :- Simplest approach with respect to other solutions
> 
> Probably… But it is still incorrect. Initialising COM on _a thread_ doesn't mean you can call COM object methods from **any thread** in your application.
> 
> As the documentation says, “You need to initialize the COM library on a thread before you call any of the library functions…” (I have this very quote in [my comment above](https://github.com/openjdk/jdk/pull/14898#discussion_r1272027291).)
> 
> > 2. Hold the call till CoInitialize : - Can end up in deadlock situation so that need to be taken care.
> 
> It's a challenge but it is the only right way, as far as I can see… Perhaps not to wait till `::CoInitialize` is called on a thread (see above) but to transfer a call for handling to a thread where COM is initialised.
> 
> > 3. Do another CoInitialize before call  :- not a good approach
> 
> Yet it does not break the rules: you can initialise COM, call an API and then call [`CoUninitialize`](https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-couninitialize). If you don't keep any references to COM objects, it is valid.

Tried this approach and it works, we may need to do the similar approach for another function `DAUDIO_GetDirectAudioDeviceDescription` because this also calling `DirectSoundEnumerateW `& `DirectSoundCaptureEnumerateW `

> 
> > _This implies that any thread which creates DirectSound objects must initialise COM first. I can't see it happening. Do I miss anything?_ `DS_StartBufferHelper` class initializing COM in thread and all member functions checking initialization status before making any calls. So another solution may be restructure the code for accessing `isInitialized()` method from outer methods and proceed. (approach 2 in better way)
> > Tried this approach with `::CoInitialize` and observed _truncated names_, but its not failing for lock unlock workflow.
> 
> As I explained above, it would still be incorrect: if enumerating sound devices requires COM, it must be initialised on the thread where you enumerate the devices.

Got your point.

Reply:
> Probably… But it is still incorrect. Initialising COM on a thread doesn't mean you can call COM object methods from any thread in your application.

Note that "Objects created on a COM thread in a multithread apartment (MTA) must be able to receive method calls from other threads at any time", so if we initialisethe devices on one thread using COM they will be avaliable on other threads as well. But the spec for DirectSoundEnumerate says nothing about COM... But from some examples I see ppl wrap it by the CoInitializeEx(NULL, COINIT_MULTITHREADED)/CoUninitialize();

Noe sure we can call CoInitializeEx on the main thread.

Reply:
> > Probably… But it is still incorrect. Initialising COM on a thread doesn't mean you can call COM object methods from any thread in your application.
> 
> Note that "Objects created on a COM thread in a multithread apartment (MTA) must be able to receive method calls from other threads at any time", so if we initialisethe devices on one thread using COM they will be avaliable on other threads as well.

But from the threads where you initialised COM as MTA.

The docs for `CoInitializeEx` clearly state you have to initialise COM before you can use it. Thus, calling methods of COM objects from an arbitrary thread in your application is wrong, isn't it?

> But the spec for DirectSoundEnumerate says nothing about COM... But from some examples I see ppl wrap it by the CoInitializeEx(NULL, COINIT_MULTITHREADED)/CoUninitialize();
> 
> Noe sure we can call CoInitializeEx on the main thread.

It's for any thread not only the main thread… You can all `AudioSystem.getMixerInfo` from any thread in a Java app, if `DirectSoundEnumerate` needs COM to be initialised even though the documentation doesn't mention it, we have to initalise COM. Thus there are two options:

1. Call `CoInitializeEx`, enumerate devices, and call `CoUninitialize`.
2. Create an executor service which initialises COM on its worker thread, enumerate devices on the worker thread and return the devices. The worker thread may terminate (after it calls `CoUninitialize`) if it doesn't receive more requests in a period of time.

Reply:
> > > 3. Do another CoInitialize before call  :- not a good approach
> > 
> > Yet it does not break the rules: you can initialise COM, call an API and then call [`CoUninitialize`](https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-couninitialize). If you don't keep any references to COM objects, it is valid.
> 
> Tried this approach and it works, we may need to do the similar approach for another function `DAUDIO_GetDirectAudioDeviceDescription` because this also calling `DirectSoundEnumerateW `& `DirectSoundCaptureEnumerateW `

Does this approach return full names for devices, without truncating them?

Reply:
But if that code needed com my expectetion would be that they call CoInitialize/CoUninitialize.

This CoInitialize initially was added here https://bugs.openjdk.org/browse/JDK-6950553 (check the stack trace),
https://github.com/openjdk/jdk/commit/b8b9c35567919f9f375cc3610d823524cf5b8582 can we check that the DAUDIO_GetDirectAudioDeviceDescription  is really used that com object w/o initialisation?
>1. Call CoInitializeEx, enumerate devices, and call CoUninitialize.
>2. Create an executor service which initialises COM on its worker thread, enumerate devices on the worker thread and return the devices. The worker thread may terminate (after it calls CoUninitialize) if it doesn't receive more requests in a period of time.

We can try reuse the "Filechooser.invokeCom" machinery, but I do not think I like it.

Reply:
is it possible that our "::CoInitialize(NULL);" prevents the ::CoInitializeEx(NULL, COINIT_MULTITHREADED); in the lib later?

Reply:
> is it possible that our "::CoInitialize(NULL);" prevents the ::CoInitializeEx(NULL, COINIT_MULTITHREADED); in the lib later?

It is not only possible, it is explicitly stated. If the threading mode is different, the function fails with `RPC_E_CHANGED_MODE`.

> But if that code requare com my expectation would be that they call CoInitialize/CoUninitialize.

Exactly! Microsoft docs don't state COM needs be initialised for enumerating devices but it looks like it needs to be.

> This CoInitialize initially was added here [JDK-6950553](https://bugs.openjdk.org/browse/JDK-6950553) (check the stack trace)

It somewhat confirms, you can't access COM without initialising it.